### PR TITLE
:mortar_board: Topic20 Add search method code to repo

### PR DESCRIPTION
**Pull request by jameshughes89**
### What
Add search method class to repo. 

### Why
So they can download it and play with it if they want. 

### Where
This will be used in Topic20 

### Additional Notes
Again, the last time this was discussed the conclusion was to not include them, but I now figure the argument to not include them is pointless --- if they don't get them from me, they will goog.

### CODE REVIEW

File: src/main/java/SearchingMethods.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+public class SearchingMethods {
```
Comment (twentylemon) [inline src/main/java/SearchingMethods.java:1]: Probably want to add simple tests for these.

**Comment by jameshughes89** [comment]
To keep binary search simple, there is no guarantee of the element being the first occurrence.

**Review by twentylemon — COMMENTED** [review]
I typically use the word _needle_ instead of _target_ when refering to the search term (and _haystack_ for the array, though _data_, _array_ or _collection_ all work). Target has a weird aggressive connotation to it (imo), whereas needle has the implication you're looking for it. That might be a me-ism though.

File: src/test/java/SearchingMethodsTest.java [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+
+    @Test
+    void iterativeBinarySearch_targetAtArbitraryIndex_returnsCorrectIndex() {
+        Integer[] myArray = {0, 4, 8, 12, 18, 25, 31, 33, 36, 35, 37, 40, 40, 40, 49, 55, 61, 62, 63, 74, 81, 89, 90, 98};
+        Integer myTarget = 55;
+        assertEquals(15, SearchingMethods.iterativeBinarySearch(myTarget, myArray));
+    }
+
+    @Test
+    void iterativeBinarySearch_unorderedArray_returnsSentinel() {
```
Comment (twentylemon) [inline src/test/java/SearchingMethodsTest.java:184]: That's not the defined behaviour for binary search in general. Searching an unordered collection may still yield a result (consider the list [3,2,1,3] and search for 3, you'll touch indexes 1, 2 then 3 and correctly return 3).

File: src/test/java/SearchingMethodsTest.java [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+
+    @Test
+    void iterativeLinearSearch_targetAtNthIndex_returnsCorrectIndex() {
+        Integer[] myArray = {0, 4, 8, 12, 18, 25, 31, 33, 36, 35, 37, 40, 40, 40, 49, 55, 61, 62, 63, 74, 81, 89, 90, 98};
+        Integer myTarget = 98;
+        assertEquals(23, SearchingMethods.iterativeLinearSearch(myTarget, myArray));
+    }
+
+    @Test
+    void iterativeLinearSearch_targetAtArbitraryIndex_returnsCorrectIndex() {
```
[the remark this answers is not in the document]
Comment (jameshughes89) [reply src/test/java/SearchingMethodsTest.java:58]: I love `needle` & `haystack`. defo switching.

File: src/test/java/SearchingMethodsTest.java (hunk 2 above)
Comment (jameshughes89) [reply src/test/java/SearchingMethodsTest.java:184]: I did take pause on this one as, well, exactly what you said. Easiest to just remove? :bomb:

Comment (twentylemon) [reply src/test/java/SearchingMethodsTest.java:184]: Imo, yes. Leaving tests out is a way to show that behaviour is undefined.

**Review by twentylemon — APPROVED** [review]

**Pull request merged**